### PR TITLE
MDEV-27467: innodb to enfore the minimum innodb_buffer_pool_size in S…

### PR DESCRIPTION
**Pull request by grooverdan**
…ET GLOBAL

.. to be the same as startup.

In resolving MDEV-27461, BUF_LRU_MIN_LEN (256) is the minium number of
pages for the innodb buffer pool size. Obviously we need more than just
flushing pages. Taking the 16k page size and its default minimum, an
extra 25% is needed on top of the flushing pages to make a workable buffer
pool.

The minimum innodb_buffer_pool_chunk_size (1M) restricts the minimum
otherwise we'd have a pool made up of different chunk sizes.

Minimum innodb buffer pool size:

Page Size, Previously minimum (startup), with change.
        4k                            5M           2M
        8k                            5M           3M
       16k                            5M           5M
       32k                           24M          10M
       64k                           24M          20M

With this patch, SET GLOBAL innodb_buffer_pool_size minimums are
enforced.

- [x] *The Jira issue number for this PR is: MDEV-27467*

## Description

A minimum buffer pool set at the startup should be the same is what is able to be dynamicly set. Otherwise issues like MDEV-27461 can be.

## How can this PR be tested?

MTR tests included. I was reluctant to add a page size combination test that just warned on the value being too low.

boundary manual tests on all 4k - 64k page sizes with different innodb buffer pool sizes where tested.

Retested MDEV-27461 case (new value < 20M so rejected).

## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [ X ] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

## Backward compatibility

We've actually lowered the minimum buffer pool size on startup. Previous dynamic changes to lower values where previously permitted, however they where dangerous and as error in this test.

### CODE REVIEW

File: storage/innobase/handler/ha_innodb.cc [hunk 1]
```diff
@@ -3639,6 +3639,16 @@ static uint innobase_partition_flags()
 	return (0);
 }
 
+/** Return the minimum buffer pool size based on page size */
+static inline ulint min_buffer_pool_size()
+{
+	ulint s= (BUF_LRU_MIN_LEN + BUF_LRU_MIN_LEN / 4) * srv_page_size;
+	/* buf_pool_chunk_size mimimum is 1M, so round up to a multiple */
```
Comment (dr-m) [inline storage/innobase/handler/ha_innodb.cc:3646]: Typo: "minimum", not "minmimum". In the commit message it is mistyped in a different way.

Furthermore, TAB should not be used for indenting new code. Please use 2 spaces instead.

File: storage/innobase/srv/srv0srv.cc [hunk 2]
```diff
@@ -244,7 +244,8 @@ UNIV_INTERN os_event_t	srv_allow_writes_event;
 
 /** copy of innodb_buffer_pool_size */
 ulint	srv_buf_pool_size;
-const ulint	srv_buf_pool_min_size	= 5 * 1024 * 1024;
+/** Minimum consistent with ha_innodb.cc:min_buffer_pool_size() for PAGE_SIZE_MIN */
+ulint	srv_buf_pool_min_size	= 2 * 1024 * 1024;
```
Comment (dr-m) [inline storage/innobase/srv/srv0srv.cc:248]: Can we avoid this indirection via a different compilation unit, and simply refer to `MYSQL_SYSVAR(buffer_pool_size)` in `ha_innodb.cc`? Similarly, `srv_buf_pool_def_size` looks unnecessary obfuscation to me. We can simply write `MYSQL_SYSVAR_LONGLONG(buffer_pool_size, …, 128LL << 20, …)` in `ha_innodb.cc`.

File: mysql-test/suite/innodb/t/innodb.opt [hunk 3]
```diff
@@ -1 +1,2 @@
---innodb-defragment=0
\ No newline at end of file
+--innodb-defragment=0
+--loose-innodb_disable_resize_buffer_pool_debug=0
```
Comment (dr-m) [inline mysql-test/suite/innodb/t/innodb.opt:2]: I do not see `have_debug.inc` in `innodb.test`. Will the test still work for non-debug builds? Perhaps it would be better to write a separate test.

File: mysql-test/suite/innodb/r/innodb.result [hunk 4]
```diff
@@ -3390,3 +3390,9 @@ Table	Op	Msg_type	Msg_text
 test.t1	check	status	OK
 ALTER TABLE t1 FORCE;
 DROP TABLE t1;
+SET GLOBAL innodb_buffer_pool_size = (256 + 64) * @@innodb_page_size -1;
+ERROR 42000: Variable 'innodb_buffer_pool_size' can't be set to the value of '5242879'
+SHOW WARNINGS;
+Level	Code	Message
+Warning	1210	innodb_buffer_pool_size must be at least 5242880 for innodb_page_size=16384
+Error	1231	Variable 'innodb_buffer_pool_size' can't be set to the value of '5242879'
```
Comment (dr-m) [inline mysql-test/suite/innodb/r/innodb.result:3398]: This test would be better added to a file that uses `innodb_page_size.inc` so that it will be tested with all 5 page sizes.

File: mysql-test/suite/innodb/t/innodb_buffer_pool_resize_with_chunks.opt [hunk 5]
```diff
@@ -1,3 +1,3 @@
 --innodb-buffer-pool-size=16M
---innodb-buffer-pool-chunk-size=2M
+--innodb-buffer-pool-chunk-size=4M
 --innodb-page-size=4k
```
Comment (dr-m) [inline mysql-test/suite/innodb/t/innodb_buffer_pool_resize_with_chunks.opt:3]: Why is this change necessary? It should be explained in the commit message if this file does not allow comments. I suppose that it is because of the reduced minimum buffer pool size and the change to the `.test` file.

File: storage/innobase/handler/ha_innodb.cc [hunk 6]
```diff
@@ -3830,12 +3840,14 @@ innobase_init(
 
 	/* The buffer pool needs to be able to accommodate enough many
 	pages, even for larger pages */
-	if (UNIV_PAGE_SIZE > UNIV_PAGE_SIZE_DEF
-	    && innobase_buffer_pool_size < (24 * 1024 * 1024)) {
+	srv_buf_pool_min_size= min_buffer_pool_size();
```
Comment (dr-m) [inline storage/innobase/handler/ha_innodb.cc:3843]: This assignment does not appear to be reflected by `mysql-test/suite/sys_vars/r/sysvars_innodb.result`, which is (unsurprisingly) showing the changed initial value of 2 megabytes when using the default page size of 16 kilobytes. I think that this assignment should better refer to `MYSQL_SYSVAR(buffer_pool_size).min_val` if that is possible. Ideally, the reported minimum value for the default `innodb_page_size=16k` would not change.

**Review by dr-m — CHANGES_REQUESTED** [review]
If we are going to change the minimum value at server startup, I think that we should try to do it in a way that does not affect the default page size (`innodb_page_size=16k`).

File: mysql-test/suite/innodb/t/innodb.opt (hunk 3 above)
Comment (grooverdan) [reply mysql-test/suite/innodb/t/innodb.opt:2]: The purpose of `loose` here is to work with debug/non-debug. Debug builds by default resist any changing of the buffer pool.

File: storage/innobase/handler/ha_innodb.cc (hunk 6 above)
Comment (grooverdan) [reply storage/innobase/handler/ha_innodb.cc:3843]: > This assignment does not appear to be reflected by `mysql-test/suite/sys_vars/r/sysvars_innodb.result`, which is (unsurprisingly) showing the changed initial value of 2 megabytes when using the default page size of 16 kilobytes. I think that this assignment should better refer to `MYSQL_SYSVAR(buffer_pool_size).min_val` if that is possible.

Doing that now. I don't think its complete.

> Ideally, the reported minimum value for the default `innodb_page_size=16k` would not change.
I don't think this is possible (unless we make 5M the absolute minimum on all page sizes). The command line args are passed in order and the resize happens before innodb code even starts.
```
$  sql/mysqld --innodb-buffer-pool-chunk-size=1M --innodb-buffer-pool-size=1M --innodb-page-size=4k  --datadir=/tmp/${PWD##*/}-datadir4
2022-01-18 19:45:37 140087854176192 [Note] sql/mysqld (mysqld 10.2.42-MariaDB) starting as process 202208 ...
2022-01-18 19:45:37 140087854176192 [Warning] option 'innodb-buffer-pool-size': signed value 1048576 adjusted to 5242880
2022-01-18 19:45:37 140087854176192 [Note] InnoDB: innodb_page_size=4096
```

File: storage/innobase/handler/ha_innodb.cc [hunk 7]
```diff
@@ -3639,6 +3639,60 @@ static uint innobase_partition_flags()
...
 }
 
+/** Return the minimum buffer pool size based on page size */
+static inline ulint min_buffer_pool_size()
+{
+  ulint s= (BUF_LRU_MIN_LEN + BUF_LRU_MIN_LEN / 4) * srv_page_size;
+  /* buf_pool_chunk_size minimum is 1M, so round up to a multiple */
+  if (s % (1 << 20))
+     s+= (1 << 20) - (s % (1 << 20));
+  return(s);
```
Comment (dr-m) [inline storage/innobase/handler/ha_innodb.cc:3649]: This could be:
```c++
  if (ulint remainder= s % (1<< 20))
    s+= (1 << 20) - remainder;
  return s;
```
However, it would be even better to use pure arithmetics without any condition:
```c++
  ulint alignment= 1U << 20;
  return UT_CALC_ALIGN(s, alignment);
```

File: storage/innobase/handler/ha_innodb.cc [hunk 8]
```diff
@@ -22653,3 +22689,21 @@ ib_push_frm_error(
...
+{
+  const ib_uint64_t	m = ((ib_uint64_t)srv_buf_pool_instances) * srv_buf_pool_chunk_unit;
+  size = ut_max((size_t) size, (size_t) MYSQL_SYSVAR_NAME(buffer_pool_size).min_val);
+
+  if (size % m == 0) {
+    return(size);
+  } else {
+    return (ulint)((size / m + 1) * m);
+  }
+}
```
Comment (dr-m) [inline storage/innobase/handler/ha_innodb.cc:22709]: We could ask why `innodb_buffer_pool_size` has been declared as `longlong` instead of `size_t` in the first place. But, I think that it was fixed in a later version already.

I would avoid old InnoDB style excessive line breaks and `{}` around single-statement blocks in new code.
```c++
ulint buf_pool_size_align(ulint size)
{
  const size_t m= …;
  size= std::max(size, size_t(MYSQL_SYSVAR_NAME(buffer_pool_size).min_val);
  return size % m
    ? size
    : (size / m + 1) * m;
}
```
Note: `UT_CALC_ALIGN()` can’t be used, because `innodb_buffer_pool_chunk_size` is not necessarily a power of 2.

File: mysql-test/suite/innodb/t/restart.test [hunk 9]
```diff
@@ -77,3 +77,28 @@ SELECT * FROM tr;
...
+  WHERE LOWER(variable_name) = 'innodb_buffer_pool_resize_status';
+
+SELECT @@innodb_buffer_pool_size INTO @innodb_buffer_pool_size_orig;
+SELECT CEILING((256 + 64) * @@innodb_page_size / 1048576) * 1048576 INTO @min_pool_size;
+--replace_regex /value of '\d+/value of 'WRONG_VALUE/
+--error ER_WRONG_VALUE_FOR_VAR
+EXECUTE IMMEDIATE 'SET GLOBAL innodb_buffer_pool_size = ?' USING (@min_pool_size -1);
+
+--replace_regex /least \d+/least MIN_VAL/ /innodb_page_size=\d+/innodb_page_size=PAGE_SIZE/ /value of '\d+/value of 'WRONG_VALUE/
+SHOW WARNINGS;
```
Comment (dr-m) [inline mysql-test/suite/innodb/t/restart.test:98]: Instead of filtering out the values, I would prefer to add `.rdiff` files for the different values so that any inadvertent change to the limits in the future will be caught by this test. The basic `.result` file should be recorded with the default page size (`./mtr --record innodb.restart,16k`).

**Review by dr-m — APPROVED** [review]
Thank you, it is better now. I wonder a little why the result of the test `sys_vars.sysvars_innodb` is still showing the minimum value as 2MiB for the default page size of 16KiB, but I think that that difference is acceptable.

This is OK to push after addressing my feedback.

File: mysql-test/suite/innodb/t/innodb.opt (hunk 3 above)
Comment (dr-m) [reply mysql-test/suite/innodb/t/innodb.opt]: Oh, I see it now. I somehow missed the `disable` in the name and thought that the purpose of the parameter was to disable the resizing, not to enable it.

I guess that this parameter was introduced in MySQL because buffer pool resizing in stress testing was revealing lots of trouble (genuine bugs that would affect buffer pool resizing while the server is under load). With some recent fixes, such as MDEV-26033, it should work more reliably.
